# MC316 Final Exam – Part B: PYTHON

* You will have 1 hour and 40 minutes to finish the test.
* This part includes 10 questions. Each question is worth 15 points.
* You get extra 30 points for full answers.
* Please read the instructions carefully. 

## I. SET UP


### Import pandas library for use in scripting. 

In [2]:
import pandas as pd

### Import the datasets and name them.
For this test, you will analyze revenue and expenditure data released by Kansas Department of Administration for the fiscal year 2018. While the original data can be directly downloaded [here](http://kanview.ks.gov/DataDownload.aspx), for your convenience, the instructor had them uploaded to a server at the following urls. 

#### For Kansas State University:
1. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv
2. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv

#### For University of Kansas:
3. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv
4. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv

You will need to import all 4 datasets and name them in an easy way to call out later.

*Please refer to the [Glossary](http://kanview.ks.gov/Glossary.aspx#f) for data description.

In [20]:
KSU = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv") 

In [21]:
Rev = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

In [23]:
KU = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")


In [30]:
data = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")
            

## II. REQUIRED QUESTIONS
* Make sure you type the answers in the cells right below the codes.
* If you need to add extra cells, click the + button in the toolbar up the top. 

### Q1: What is the total revenue recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total revenue recorded for University of Kansas in the same year?

In [71]:
Rev.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1910 entries, 0 to 1909
Data columns (total 22 columns):
Fiscal_Year                        1910 non-null int64
Fiscal_Quarter                     1910 non-null int64
Business_Unit                      1910 non-null int64
Business_Unit_Descr                1910 non-null object
GAAP_Fund_Type                     1910 non-null int64
GAAP_Fund_Type_Descr               1910 non-null object
Fund_Code                          1910 non-null int64
Fund_Code_Descr                    1910 non-null object
Budget_Ref                         1910 non-null int64
Budget_Ref_Descr                   1910 non-null object
Program_Code                       1910 non-null int64
Program_Code_Descr                 1910 non-null object
State_Function                     1910 non-null int64
State_Function_Descr               1910 non-null object
Primary_Account_Code               1910 non-null int64
Primary_Account_Code_Descr         1910 non-null object
Interm

In [70]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1077 entries, 0 to 1076
Data columns (total 22 columns):
Fiscal_Year                        1077 non-null int64
Fiscal_Quarter                     1077 non-null int64
Business_Unit                      1077 non-null int64
Business_Unit_Descr                1077 non-null object
GAAP_Fund_Type                     1077 non-null int64
GAAP_Fund_Type_Descr               1077 non-null object
Fund_Code                          1077 non-null int64
Fund_Code_Descr                    1077 non-null object
Budget_Ref                         1077 non-null int64
Budget_Ref_Descr                   1077 non-null object
Program_Code                       1077 non-null int64
Program_Code_Descr                 1077 non-null object
State_Function                     1077 non-null int64
State_Function_Descr               1077 non-null object
Primary_Account_Code               1077 non-null int64
Primary_Account_Code_Descr         1077 non-null object
Interm

In [108]:
Rev.Total.sum()

854324643.84

In [109]:
data.Total.sum()

808767902.55

In [110]:
854324643.84-808767902.55

45556741.29000008

<b>Answer:</b> K-State made this much more in revenue than KU 45556741.29000008

### Q2: What is the total expense recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total expense recorded for University of Kansas in the same year?

In [106]:
KSU.Total.sum()

1018584373.02

In [111]:
KU.Total.sum()

990238875.6099999

In [112]:
1018584373.02-990238875.6099999

28345497.410000086

<b>Answer:</b> K-State had 28345497.410000086 in expenses compared to KU


### Q3: Where did the two universities' primary revenues come from? 
*Refer to the description [here](http://kanview.ks.gov/Revenues/AcctCodeRevenues.aspx) to write your answer.

In [200]:
Rev.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   637
REV FROM USE OF MONEY & PROP      354
REVENUE TRANSFERS                 341
NON REVENUE RECEIPTS              259
OTHER REVENUE                     160
GIFT DONATION FEDERAL GRANTS      152
TAXES                               7
Name: Primary_Account_Code_Descr, dtype: int64

In [201]:
data.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   376
REV FROM USE OF MONEY & PROP      251
NON REVENUE RECEIPTS              157
REVENUE TRANSFERS                 153
OTHER REVENUE                      57
TAXES                              45
GIFT DONATION FEDERAL GRANTS       38
Name: Primary_Account_Code_Descr, dtype: int64

<b>Answer:</b> Agency earnings

### Q4: What was the biggest source of revenues for Kansas State University? Compare to the one of University of Kansas.

Remove hashtag and run the following code to suppress scientific notation:

In [204]:
Rev.groupby(Primary_Account_Code_Descr).values_counts()

NameError: name 'Primary_Account_Code_Descr' is not defined

<b>Answer:</b> 

### Q5: What accounts contributed to Kansas State University's biggest source of revenues?
If you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [2]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

After cleaning the spaces, you can start filtering your datasets to find the answer for Q5.

<b>Answer:</b>

### Q6: Where did the money go? (Answer for both universities)
*Refer to the description [here] (http://kanview.ks.gov/Expenditures/AcctCodeExpenditures.aspx) to write your answer.

<b>Answer:</b> 

### Q7: What costs two universities the most? 

<b>Answer:</b> 

### Q8: Calculate the percentage of this cost over each university's total expense.

<b>Answer:</b> 

### Q9: Breakdown the cost you found in Q7 for both universities.
Again, if you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [3]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

<b>Answer:</b>

### Q10: Compare the breakdown costs from Q9 side by side.
When you merge the two subsets, make sure to add the attribute <b>how ="outer"</b> so that you can keep every item from both sets.
*For example: merge = pd.merge(df1, df2, how="outer", on="columnname")

<b>Answer:</b>

## III. BONUS QUESTIONS
* Each question's worth 15 points.


### B1: Which university made a profit in the fiscal year of 2018?

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1910 entries, 0 to 1909
Data columns (total 22 columns):
Fiscal_Year                        1910 non-null int64
Fiscal_Quarter                     1910 non-null int64
Business_Unit                      1910 non-null int64
Business_Unit_Descr                1910 non-null object
GAAP_Fund_Type                     1910 non-null int64
GAAP_Fund_Type_Descr               1910 non-null object
Fund_Code                          1910 non-null int64
Fund_Code_Descr                    1910 non-null object
Budget_Ref                         1910 non-null int64
Budget_Ref_Descr                   1910 non-null object
Program_Code                       1910 non-null int64
Program_Code_Descr                 1910 non-null object
State_Function                     1910 non-null int64
State_Function_Descr               1910 non-null object
Primary_Account_Code               1910 non-null int64
Primary_Account_Code_Descr         1910 non-null object
Interm

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1077 entries, 0 to 1076
Data columns (total 22 columns):
Fiscal_Year                        1077 non-null int64
Fiscal_Quarter                     1077 non-null int64
Business_Unit                      1077 non-null int64
Business_Unit_Descr                1077 non-null object
GAAP_Fund_Type                     1077 non-null int64
GAAP_Fund_Type_Descr               1077 non-null object
Fund_Code                          1077 non-null int64
Fund_Code_Descr                    1077 non-null object
Budget_Ref                         1077 non-null int64
Budget_Ref_Descr                   1077 non-null object
Program_Code                       1077 non-null int64
Program_Code_Descr                 1077 non-null object
State_Function                     1077 non-null int64
State_Function_Descr               1077 non-null object
Primary_Account_Code               1077 non-null int64
Primary_Account_Code_Descr         1077 non-null object
Interm

<b>Answer:</b> Both universities made a profit in the fiscal year of 2018

### B2: Which software program do you like the most in this course? Why?

<b>Answer:</b> I like python because the tutorial is easier to understand than R Studio

### END OF THE TEST